api/src/main/java/org/eclipse/microprofile/config/ConfigAccessor.java

```diff
@@ -0,0 +1,174 @@
+     * property name.
+     *
+     * @param propertyName the new property name (must not be {@code null})
+     * @return the derived accessor
+     */
+    ConfigAccessor<T> forPropertyName(String propertyName);
+
+    /**
+     * Get the associated configuration value.
+     *
+     * @return the configuration value
+     *
+     * @throws IllegalArgumentException if the conversion of the value or a string default value failed
+     * @throws NoSuchElementException if the configuration value converts to an empty value, or the
+     *      configuration value is not given and no default value is set, or the configuration value is not given
+     *      and a string default value was set which converts to an empty value
+     * @throws IllegalStateException if the accessor is not associated with a type or converter
+     */
+    T getValue() throws IllegalArgumentException, NoSuchElementException, IllegalStateException;
+
+    /**
+     * Get a derived accessor which returns an optional value.  The resultant accessor cannot be used to derive
+     * accessors with a different type or converter, or type-specific default value.
+     *
+     * @return the derived accessor
+     * @throws IllegalStateException if the accessor is already optional, or if a type or converter is not
+     *      associated with this accessor
+     */
+    ConfigAccessor<Optional<T>> optional() throws IllegalStateException;
+
+    /**
+     * Get a derived accessor which yields values converted by the given converter.  If this method is called on
+     * an accessor that is already associated with a type or converter, the previous type or converter is discarded
+     * and the new accessor will not reference it.
+     *
+     * @param converter the converter to use (must not be {@code null})
+     * @param <U> the target type
+     * @return the derived accessor
+     * @throws IllegalStateException if the accessor is optional, or if the accessor is already associated with a
+     *      type-specific default value
+     */
+    <U> ConfigAccessor<U> convertedWith(Converter<U> converter) throws IllegalStateException;
+
+    /**
+     * Get a derived accessor which yields values converted by the default converter for the given type.  If this method
+     * is called on an accessor that is already associated with a type or converter, the previous type or converter is
+     * discarded and the new accessor will not reference it.
+     *
+     * @param clazz the type of the value to return (must not be {@code null})
+     * @param <U> the target type
+     * @return the derived accessor
+     * @throws IllegalStateException if the accessor is optional, or if the accessor is already associated with a
+     *      type-specific default value, or if there is no converter available for the given type
+     */
+    <U> ConfigAccessor<U> convertedForType(Class<U> clazz) throws IllegalStateException;
```

Comment:
I think this pattern leads to the creation of ConfigAccessor being mutable or thrown away ConfigAccessor. This is why we introduced the builder pattern in the first place.

Reply:
I would expect a properly tuned JDK to elide the vast majority of these allocations or else detect that they do not escape.  The config accessor is not mutable (by specification) so I'm not sure what you mean by saying it's mutable.

The "fluent" style of API is widely used already.  Using a builder on the other hand is much more unwieldy from a user's perspective, for something that will be thrown away immediately.

Reply:
Here are some other things to consider:

* Using the fluent immutable API means the result is perfectly type safe.  The builder approach means that no mutations are allowed which would change the type of the result (otherwise ugly tricks like unsafely re-typing the object would have to be employed).  If you forbid re-typing, the type or converter must be passed in to `access` and cannot be changed later; this also means the accessor result must have separate methods for every result type (for example, optional values would need their own method, as would collection values (if those were ever added), etc.).
* Using the builder API, you end up with an extra, strange `build` step.  These two things together make the API more clunky and decrease its flexibility: `config.access("some.prop", Foo.class).build().getOptionalValue()` versus `config.access("some.prop").withType(Foo.class).optional().getValue()`.  It's not entirely obvious that to the user that calling `build` is necessary, whereas with the immutable option, IDE completion would make it clear what operations are possible in a fairly unambiguous way based on the result type and possible method completions.
* The fluent immutable API also reflects the immutability of the `config` object itself (not reflecting the mutability of the backing data, but the actual configuration object and *its* configuration from an API perspective).

Reply:
I agree with David, this pattern is fine and make things type safe and consistent.

```
cs = config.access("some.prop").withType(Foo.class).WithDefault("blah")
cs.getValue() -> "blah"
cs.WithDefault("Yes!") // <- this does not mutate the cs instance
cs.getValue() -> "blah"

// If users want to provide another default, he has to take the returned ConfigAccessor:
cs.withDefault("Yes!").getValue() -> "Yes!"
```

In most cases, users won't even keep a reference on the `ConfigAccessor` and will directly take the returned value for `getValue()`.

Reply:
It's very similar to the original design and mixes up builder and ConfigAccessor. It's basically the old ConfigValue. We have been there over a year ago.

Reply:
I prefer the builder pattern, which was in the master before. With the build pattern, the usage will be a functional style. You build one configAccessor and use it.